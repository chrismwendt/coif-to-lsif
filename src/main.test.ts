import { index } from './index'
import { Edge, Vertex } from 'lsif-protocol'
import _ from 'lodash'
import * as path from 'path'
import * as cp from 'child_process'

const GENERATE = false

function generate(example: string): void {
    cp.execFileSync('./generate-csv', ['$CXX -c *.cpp'], {
        env: {
            ABSROOTDIR: path.resolve(`examples/${example}/root`),
            ABSOUTDIR: path.resolve(`examples/${example}/output`),
            CLEAN: 'true',
        },
    })
}

async function indexExample(example: string): Promise<(Edge | Vertex)[]> {
    if (GENERATE) {
        generate(example)
    }

    const output: (Edge | Vertex)[] = []

    await index({
        csvFileGlob: `examples/${example}/output/*.csv`,
        root: `examples/${example}/root`,
        emit: item =>
            new Promise(resolve => {
                output.push(item)
                resolve()
            }),
    })

    return output
}

test('does not emit items with duplicate IDs', async () => {
    const output = await indexExample('five')

    const setsOfDupes = _(output)
        .groupBy(item => item.id)
        .values()
        .map(group => ({ group, count: group.length }))
        .value()
        .filter(({ count }) => count > 1)
        .map(({ group }) => group)

    if (setsOfDupes.length > 0) {
        fail(
            new Error(
                `Sets of lines with duplicate IDs:\n` +
                    setsOfDupes
                        .map(dupes =>
                            dupes.map(item => JSON.stringify(item)).join('\n')
                        )
                        .join('\n\n')
            )
        )
    }
})

test('five', async () => {
    const output = (await indexExample('five')).map(v => JSON.stringify(v))

    expect(output.join('\n')).toMatchSnapshot()
})

test('cross-app', async () => {
    const app = (await indexExample('cross-app')).map(v => JSON.stringify(v))
    expect(app.join('\n')).toMatchSnapshot()
})

test('cross-lib', async () => {
    const lib = (await indexExample('cross-lib')).map(v => JSON.stringify(v))
    expect(lib.join('\n')).toMatchSnapshot()
})
