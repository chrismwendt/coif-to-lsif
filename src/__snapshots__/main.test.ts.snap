// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one 1`] = `
"{\\"id\\":\\"meta\\",\\"type\\":\\"vertex\\",\\"label\\":\\"metaData\\",\\"projectRoot\\":\\"file:///\\",\\"version\\":\\"0.4.0\\",\\"positionEncoding\\":\\"utf-16\\",\\"toolInfo\\":{\\"name\\":\\"lsif-cpp\\",\\"args\\":[\\"examples/one/*.lsif.in\\",\\"examples/one\\"],\\"version\\":\\"dev\\"}}
{\\"id\\":\\"project\\",\\"type\\":\\"vertex\\",\\"label\\":\\"project\\",\\"kind\\":\\"cpp\\"}
{\\"id\\":\\"projectBegin\\",\\"data\\":\\"project\\",\\"type\\":\\"vertex\\",\\"label\\":\\"$event\\",\\"kind\\":\\"begin\\",\\"scope\\":\\"project\\"}
{\\"id\\":\\"document:one\\",\\"type\\":\\"vertex\\",\\"label\\":\\"document\\",\\"uri\\":\\"file:///one\\",\\"languageId\\":\\"cpp\\",\\"contents\\":\\"\\"}
{\\"id\\":\\"documentBegin:one\\",\\"data\\":\\"document:one\\",\\"type\\":\\"vertex\\",\\"label\\":\\"$event\\",\\"kind\\":\\"begin\\",\\"scope\\":\\"document\\"}
{\\"id\\":\\"one:35:33\\",\\"type\\":\\"vertex\\",\\"label\\":\\"range\\",\\"start\\":{\\"line\\":35,\\"character\\":33},\\"end\\":{\\"line\\":35,\\"character\\":49}}
{\\"id\\":\\"one:8:6\\",\\"type\\":\\"vertex\\",\\"label\\":\\"range\\",\\"start\\":{\\"line\\":8,\\"character\\":6},\\"end\\":{\\"line\\":8,\\"character\\":6}}
{\\"id\\":\\"resultSet:one:8:6\\",\\"label\\":\\"resultSet\\",\\"type\\":\\"vertex\\"}
{\\"id\\":\\"next:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"next\\",\\"outV\\":\\"one:8:6\\",\\"inV\\":\\"resultSet:one:8:6\\"}
{\\"id\\":\\"definition:one:8:6\\",\\"label\\":\\"definitionResult\\",\\"type\\":\\"vertex\\"}
{\\"id\\":\\"textDocument/definition:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"textDocument/definition\\",\\"outV\\":\\"resultSet:one:8:6\\",\\"inV\\":\\"definition:one:8:6\\"}
{\\"id\\":\\"item:textDocument/definition:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"item\\",\\"outV\\":\\"definition:one:8:6\\",\\"inVs\\":[\\"one:8:6\\"],\\"document\\":\\"document:one\\"}
{\\"id\\":\\"hover:one:8:6\\",\\"label\\":\\"hoverResult\\",\\"type\\":\\"vertex\\",\\"result\\":{\\"contents\\":\\"RGB6.t -> int\\"}}
{\\"id\\":\\"textDocument/hover:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"textDocument/hover\\",\\"outV\\":\\"resultSet:one:8:6\\",\\"inV\\":\\"hover:one:8:6\\"}
{\\"id\\":\\"reference:one:8:6\\",\\"label\\":\\"referenceResult\\",\\"type\\":\\"vertex\\"}
{\\"id\\":\\"textDocument/references:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"textDocument/references\\",\\"outV\\":\\"resultSet:one:8:6\\",\\"inV\\":\\"reference:one:8:6\\"}
{\\"id\\":\\"item:textDocument/references:definitions:one:8:6\\",\\"type\\":\\"edge\\",\\"label\\":\\"item\\",\\"outV\\":\\"reference:one:8:6\\",\\"inVs\\":[\\"one:8:6\\"],\\"property\\":\\"definitions\\",\\"document\\":\\"document:one\\"}
{\\"id\\":\\"next:one:35:33\\",\\"type\\":\\"edge\\",\\"label\\":\\"next\\",\\"outV\\":\\"one:35:33\\",\\"inV\\":\\"resultSet:one:8:6\\"}
{\\"id\\":\\"item:textDocument/references:references:one:8:6:one\\",\\"type\\":\\"edge\\",\\"label\\":\\"item\\",\\"outV\\":\\"reference:one:8:6\\",\\"inVs\\":[\\"one:35:33\\"],\\"property\\":\\"references\\",\\"document\\":\\"document:one\\"}
{\\"id\\":\\"contains:one\\",\\"type\\":\\"edge\\",\\"label\\":\\"contains\\",\\"outV\\":\\"document:one\\",\\"inVs\\":[\\"one:35:33\\",\\"one:8:6\\"]}
{\\"id\\":\\"documentEnd:one\\",\\"data\\":\\"document:one\\",\\"type\\":\\"vertex\\",\\"label\\":\\"$event\\",\\"kind\\":\\"end\\",\\"scope\\":\\"document\\"}
{\\"id\\":\\"projectContains\\",\\"type\\":\\"edge\\",\\"label\\":\\"contains\\",\\"outV\\":\\"project\\",\\"inVs\\":[\\"document:one\\"]}
{\\"id\\":\\"projectEnd\\",\\"data\\":\\"project\\",\\"type\\":\\"vertex\\",\\"label\\":\\"$event\\",\\"kind\\":\\"end\\",\\"scope\\":\\"project\\"}"
`;
